Add default values for actions documentation

revolter opened the pull request:
<!-- Thanks for contributing to _fastlane_! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] -->

### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

Related to #435.

Adds the default value of actions to their doc pages.

<!--- If it fixes an open issue, please link to the issue here. -->
<!--- Please describe in detail how you tested your changes. --->

### Description
<!--- Describe your changes in detail -->

- [x] Add `default_value_dynamic` flag to the rest of the `ConfigItem`s.

or

- [ ] <strike>Find a way to automatically detect if the `default_value` is dynamic or not. Maybe this could help:</strike>

https://github.com/fastlane/fastlane/blob/2fd9429916348d5b0ce9db3279a2ada3dea84c0d/fastlane_core/lib/fastlane_core/configuration/configuration.rb#L156-L194

It looks like this is only for `*file`s (e.g. `Deliverfile`, `Gymfile`).

revolter committed:
Add default value for actions doc pages

KrauseFx commented:
Hey @revolter, thanks for your contribution, and migrating your PR over. One thing I'm a little worried about is, that many default values are being generated on demand, based on your system environment. So the extra column might give the wrong impression of other config items having no default values, even though they might do. I'm not sure if there is a way we can solve this?

revolter commented:
Firstly, it looks exactly the same if you run `fastlane action gym` too, so the problem already existed long before. So we could solve that too if we're at it (I'm not pointing fingers or something).

Secondly, what about changing:

Key | Description | Default
----|-------------|--------
x | first | true
y | second |
z | third | false

to:

Key | Description | Default
----|-------------|--------
x | first | true
y | second | false *
z | third | false

`* = this default value is dependent on the user's system`

janpio commented:
> many default values are being generated on demand, based on your system environment.

What is this about? Examples?

revolter commented:
@janpio, `fastlane action cert`'s `keychain_path` has a default value of `/Users/revolt/Library/Keychains/login.keychain-db` for me.

janpio commented:
Hmm, outputting this might not be the most easy to understand:
https://github.com/fastlane/fastlane/blob/master/cert/lib/cert/options.rb#L53

revolter commented:
I don't understand exactly what do you mean. And, why is there an array if only the last element is selected?

janpio commented:
This is how the `default_value` is defined in the source of `cert`, see the link under the code for context. This is what would be available for docs generation.
Why it is this way, especially the `.last`? No idea.

revolter commented:
I understand, I don't know how can I add these dynamic values :-?

revolter commented:
I understood this better, and, I am proposing this solution:

For docs.fastlane.tools, use:

Key | Description | Default
----|-------------|--------
a | first | true
b | second | *
c | third | false
d | fourth | -

_- = this parameter doesn't have a default value_
_* = this default value is dependent on the user's system_

For `terminal`, use:

Key | Description | Default
----|-------------|--------
a | first | true
b | second | false*
c | third | false
d | fourth | -

_- = this parameter doesn't have a default value_
_* = this default value is dependent on the user's system_

This way, the user that reads the online docs knows what is the default value of a parameter, which parameters don't have a default value and which have a dynamic default value. And for the user that runs `fastlane action <action_name>`, we also show the evaluated dynamic default value while also letting him/she know that it won't be the same on other systems.

What do you think?

janpio commented:
Sounds good to me.

Basically: 
1. If it has a defined value, say that.
2. If it has some code that has to be evaluated, say so but nothing else.
3. If no value, say that.

#2 could link to the line in the file on Github maybe so a user can click it to find out how the default is calculated?

revolter commented:
Links to specific lines get easily desynchronized, one solution would be to link to the entire file.

janpio commented:
You know the revision when you build the docs. Just link to that version of the file with the line at that revision.

https://github.com/fastlane/fastlane/blob/f410f2a6d9d20ae8e6747db978cae7142a03d781/Dangerfile#L5 instead of https://github.com/fastlane/fastlane/blob/master/Dangerfile#L5

revolter commented:
This indeed solves one desynchronization case, but what about when someone adds a new option at the beginning of the file? All the line numbers should be updated, but most probably the respective person would forget to do it.

janpio commented:
Docs are generated automatically by a script. This script would link the "complex" default values automatically to the correct line that it already used to get the param name, param description and param default. If someone edits an action, docs are regenerated and links are updated.

revolter commented:
The `options.rb` files are not parsed, but evaluated and run like any regular ruby file, so you don't have access to that kind of info.

janpio commented:
Oh ok. 

Hmm, is there a "give me a code behind this variable" instead of "give me the value of this variable" maybe?

revolter commented:
Most probably not, but it's ok, a link to the file would suffice.

janpio commented:
Hey @revolter, could you rebase this onto current `master` so we get the new CI checks here?

Also, how would we best check if this generates a useful version of the docs to review? I still like this idea a lot and would like to get this merged or at least progressed in any way.

revolter commented:
Will do in half a day.

I will need to re-read the comments here because I kind of forgot the implications and cons of the current changes.

> I still like this idea a lot and would like to get this merged or at least progressed in any way.

Me too, I didn't forget it.

revolter commented:
I think the problem was caused by default values like:

https://github.com/fastlane/fastlane/blob/8a2e328892c34ff5d1d450a759a480cb325c5008/gym/lib/gym/options.rb#L166

which would **probably** cause the docs to actually show `#{FastlaneCore::Helper.buildlog_path}/gym`, but I'm not sure. I'll test it after rebase.

janpio commented:
Yeah, I think it might be worth to actually merge the current state if it isn't to bad and work on improvements later on. That's why I was asking how to best test the code. Ping me when you rebased.

revolter commented:
Stubborn bot 😂 I'll prefer to do it all in this PR. Also, I didn't have time last night to work on this, will let you know asap.

revolter commented:
@janpio, Rebased. But I have no idea how to test the docs.

revolter commented:
So, that line actually evaluates to `~/Library/Logs/gym`, but we wanted for the user to know that this value is for his/her system, and might be different on other systems.

revolter committed:
Add default_value_dynamic property to ConfigItem

revolter committed:
Separate dynamic default values from static ones

- if the option doesn't have a default value, print "-"
- if the option's default value is static (a literal value), print
  it directly
- if the option's default value is dynamic (a variable or a
  function call), print "*"

revolter committed:
Add draft dynamic default value for one option

revolter committed:
Separate dynamic default values from static ones in terminal

revolter commented:
I managed to create a draft implementation of [this comment](https://github.com/fastlane/fastlane/pull/10596#issuecomment-349299901):

![screen shot 2018-02-15 at 23 21 35](https://user-images.githubusercontent.com/5748627/36282187-0c007d1c-12a9-11e8-9bdd-09be526eb57c.png)

![screen shot 2018-02-15 at 23 46 57](https://user-images.githubusercontent.com/5748627/36282694-a7aedf32-12aa-11e8-90c5-2047ad8b371f.png)

revolter committed:
Fix RuboCop issues

revolter commented:
Off topic, this looks like it's an important issue:

```
#!/bin/bash --login -eo pipefail
bundle exec danger || echo "danger failed"
^D^D
Could not set up API to Code Review site for Danger

For your GitHub repo, you need to expose: DANGER_GITHUB_API_TOKEN
You may also need: DANGER_GITHUB_HOST, DANGER_GITHUB_API_BASE_URL

Found these keys in your ENV: CIRCLE_WORKFLOW_JOB_ID, CIRCLE_BUILD_NUM, CIRCLE_PR_REPONAME, CIRCLE_NODE_TOTAL, GEM_HOME, SHELL, CI_PULL_REQUESTS, SSH_CLIENT, TMPDIR, CIRCLE_INTERNAL_SCRATCH, CIRCLE_WORKFLOW_UPSTREAM_JOB_IDS, CIRCLE_PROJECT_USERNAME, SSH_TTY, LC_ALL, CIRCLE_WORKFLOW_CONCURRENCY_MAP, CIRCLE_INTERNAL_TASK_DATA, USER, CIRCLE_NODE_INDEX, CIRCLE_COMPARE_URL, CIRCLE_BRANCH, SSH_AUTH_SOCK, CIRCLE_PULL_REQUEST, CIRCLE_JOB, CIRCLE_PULL_REQUESTS, CI_PULL_REQUEST, CIRCLE_WORKING_DIRECTORY, CIRCLE_PR_USERNAME, PATH, MAIL, CIRCLE_WORKFLOW_WORKSPACE_ID, CIRCLE_USERNAME, CIRCLE_PREVIOUS_BUILD_NUM, PWD, CIRCLE_STAGE, LANG, CIRCLE_PR_NUMBER, RUBY_ENGINE, CIRCLE_BUILD_URL, CIRCLE_WORKFLOW_ID, CIRCLECI, HOME, SHLVL, GEM_ROOT, CIRCLE_SHA1, CI, CIRCLE_REPOSITORY_URL, CIRCLE_INTERNAL_CONFIG, BASH_ENV, CIRCLE_SHELL_ENV, LOGNAME, GEM_PATH, SSH_CONNECTION, LC_CTYPE, RUBY_ROOT, RUBYOPT, CIRCLE_TEST_REPORTS, CIRCLE_PROJECT_REPONAME, RUBY_VERSION, _, __CF_USER_TEXT_ENCODING, BUNDLER_ORIG_BUNDLE_BIN_PATH, BUNDLER_ORIG_BUNDLE_GEMFILE, BUNDLER_ORIG_BUNDLER_ORIG_MANPATH, BUNDLER_ORIG_BUNDLER_VERSION, BUNDLER_ORIG_GEM_HOME, BUNDLER_ORIG_GEM_PATH, BUNDLER_ORIG_MANPATH, BUNDLER_ORIG_PATH, BUNDLER_ORIG_RB_USER_INSTALL, BUNDLER_ORIG_RUBYLIB, BUNDLE_BIN_PATH, BUNDLE_GEMFILE, BUNDLER_VERSION, BUNDLER_ORIG_RUBYOPT, RUBYLIB, MANPATH.

Failing the build, Danger cannot run without API access.
You can see more information at http://danger.systems/guides/getting_started.html
danger failed
```

revolter committed:
Make sure that empty dynamic values are marked correctly

revolter committed:
Add dynamic default value state for the rest of the options

revolter committed:
Fix RuboCop issue

Resolving instead of ignoring this would require a couple more
lines with no advantage.

revolter committed:
Fix RuboCop issues

janpio commented on fastlane/lib/fastlane/actions/register_devices.rb:
```diff
@@ -75,6 +75,7 @@ def self.available_options
                                      env_name: "REGISTER_DEVICES_TEAM_ID",
                                      code_gen_sensitive: true,
                                      default_value: CredentialsManager::AppfileConfig.try_fetch_value(:team_id),
+                                       default_value_dynamic: true,
```
wrong indentation level

revolter replied:
The other lines are indented wrongly, but I didn't want to mix different changes. So I planned to do another PR that fixes them.

janpio replied:
ah ok

janpio commented:
Nice.

Ideas for future PRs:
- Parse the files and recognize dynamically if the value is static or dynamic
- Parse the files and output the dynamic value in the docs
- Parse the files, parse the dynamic value and transform it into a more readable "explanation" (`ENV[foo]` => "Environment variable `FOO`")
- Build step / rubocop that makes sure all dynamic default values are marked as such
- Replace the `dynamic=true` with `dynamic_explanation="Environment variable `FOO` or `false` if unset"

revolter commented:
I think that programmatically checking if a value is dynamic or not would be quite hard.

For example, this:

https://github.com/fastlane/fastlane/blob/487f1dae11b79f50afcd50cfee079ebb5f1b8cea/screengrab/lib/screengrab/options.rb#L43

and

https://github.com/fastlane/fastlane/blob/487f1dae11b79f50afcd50cfee079ebb5f1b8cea/fastlane/lib/fastlane/actions/flock.rb#L56

look the same, but the first one is defined like:

https://github.com/fastlane/fastlane/blob/487f1dae11b79f50afcd50cfee079ebb5f1b8cea/screengrab/lib/screengrab/options.rb#L11

which is dynamic, while the second one is defined like:

https://github.com/fastlane/fastlane/blob/487f1dae11b79f50afcd50cfee079ebb5f1b8cea/fastlane/lib/fastlane/actions/flock.rb#L4

which is static.

janpio commented:
Writing this yourself, yes. But if there is a library one could use (or even a meta feature of ruby maybe) it would be simple. But there is definitely no naive and obvious implementation, yes.
